File: backend/app/utils/logger.py
"""
Structured Logging Utility

Provides consistent, structured logging across the backend application.
Replaces scattered print() calls with a centralized logging system.

Benefits:
- Consistent log format across the app
- Log levels for filtering (DEBUG, INFO, WARNING, ERROR)
- Structured data for easier debugging
- Can be configured per environment
- Easy to integrate with log aggregation services
- Performance insights with timestamps
- Context-rich error reporting

Usage:
    from app.utils.logger import logger

    logger.info("User logged in", extra={"user_id": user_id, "email": email})
    logger.error("API call failed", extra={"endpoint": "/api/calendars", "error": str(error)})
    logger.debug("Processing video", extra={"video_id": video_id, "size": file_size})
"""

import logging
import sys
from typing import Any, Optional
from datetime import datetime
import json


class StructuredFormatter(logging.Formatter):
    """
    Custom formatter that outputs structured JSON logs.
    Makes it easier to parse and analyze logs programmatically.
    """

    def format(self, record: logging.LogRecord) -> str:
        """Format log record as structured JSON"""
        log_data = {
            'timestamp': datetime.utcnow().isoformat() + 'Z',
            'level': record.levelname,
            'message': record.getMessage(),
            'module': record.module,
            'function': record.funcName,
            'line': record.lineno,
        }

        # Add extra fields if present
        if hasattr(record, 'extra') and record.extra:
            log_data['data'] = record.extra

        # Add exception info if present
        if record.exc_info:
            log_data['exception'] = self.formatException(record.exc_info)

        return json.dumps(log_data)


class ColoredFormatter(logging.Formatter):
    """
    Custom formatter with colors for console output.
    Makes logs easier to read during development.
    """

    # ANSI color codes
    COLORS = {
        'DEBUG': '\033[36m',      # Cyan
        'INFO': '\033[32m',       # Green
        'WARNING': '\033[33m',    # Yellow
        'ERROR': '\033[31m',      # Red
        'CRITICAL': '\033[35m',   # Magenta
        'RESET': '\033[0m'        # Reset
    }

    def format(self, record: logging.LogRecord) -> str:
        """Format log record with colors"""
        color = self.COLORS.get(record.levelname, self.COLORS['RESET'])
        reset = self.COLORS['RESET']

        # Format timestamp
        timestamp = datetime.utcnow().strftime('%Y-%m-%d %H:%M:%S')

        # Build log message
        log_parts = [
            f"{timestamp}",
            f"{color}{record.levelname:8}{reset}",
            f"{record.module}.{record.funcName}:{record.lineno}",
            f"{record.getMessage()}"
        ]

        message = " | ".join(log_parts)

        # Add extra data if present
        if hasattr(record, 'extra') and record.extra:
            message += f"\n  Data: {json.dumps(record.extra, indent=2)}"

        # Add exception info if present
        if record.exc_info:
            message += f"\n{self.formatException(record.exc_info)}"

        return message


class CustomAdapter(logging.LoggerAdapter):
    """
    Custom logger adapter that allows passing structured data.
    Makes it easy to add context to log messages.
    """

    def process(self, msg: str, kwargs: dict) -> tuple:
        """Process log message and extract extra data"""
        extra = kwargs.pop('extra', {})
        if 'extra' not in kwargs:
            kwargs['extra'] = {}
        kwargs['extra']['extra'] = extra
        return msg, kwargs


def setup_logger(
    name: str = 'advent_calendar',
    level: int = logging.INFO,
    use_json: bool = False
) -> logging.Logger:
    """
    Set up and configure logger with appropriate handlers.

    Args:
        name: Logger name
        level: Logging level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
        use_json: Whether to use JSON formatting (useful for production)

    Returns:
        Configured logger instance
    """
    logger = logging.getLogger(name)
    logger.setLevel(level)

    # Remove existing handlers to avoid duplicates
    logger.handlers.clear()

    # Create console handler
    console_handler = logging.StreamHandler(sys.stdout)
    console_handler.setLevel(level)

    # Set formatter based on environment
    if use_json:
        formatter = StructuredFormatter()
    else:
        formatter = ColoredFormatter()

    console_handler.setFormatter(formatter)
    logger.addHandler(console_handler)

    # Prevent propagation to avoid duplicate logs
    logger.propagate = False

    return logger


# Create default logger instance
# Use INFO level by default, can be configured via environment variable
import os
log_level_str = os.getenv('LOG_LEVEL', 'INFO').upper()
log_level = getattr(logging, log_level_str, logging.INFO)
use_json_logs = os.getenv('LOG_FORMAT', 'colored').lower() == 'json'

_logger = setup_logger(level=log_level, use_json=use_json_logs)
logger = CustomAdapter(_logger, {})


# Convenience functions for common logging patterns

def log_request(method: str, endpoint: str, status_code: int, duration_ms: float):
    """Log HTTP request with structured data"""
    logger.info(
        f"{method} {endpoint} - {status_code}",
        extra={
            'method': method,
            'endpoint': endpoint,
            'status_code': status_code,
            'duration_ms': duration_ms
        }
    )


def log_error(error: Exception, context: Optional[dict] = None):
    """Log error with exception details"""
    logger.error(
        f"Error occurred: {str(error)}",
        exc_info=True,
        extra=context or {}
    )


def log_api_call(endpoint: str, method: str, success: bool, duration_ms: float = 0):
    """Log API call with details"""
    level = logger.info if success else logger.error
    level(
        f"API call: {method} {endpoint}",
        extra={
            'endpoint': endpoint,
            'method': method,
            'success': success,
            'duration_ms': duration_ms
        }
    )


def log_video_processing(video_id: str, action: str, status: str, details: Optional[dict] = None):
    """Log video processing events"""
    logger.info(
        f"Video {action}: {video_id} - {status}",
        extra={
            'video_id': video_id,
            'action': action,
            'status': status,
            **(details or {})
        }
    )


def log_auth_event(event: str, user_id: Optional[str] = None, email: Optional[str] = None, success: bool = True):
    """Log authentication events"""
    level = logger.info if success else logger.warning
    level(
        f"Auth event: {event}",
        extra={
            'event': event,
            'user_id': user_id,
            'email': email,
            'success': success
        }
    )

File: frontend/src/utils/logger.ts
/**
 * Structured Logging Utility
 *
 * Provides consistent, structured logging across the frontend application.
 * Replaces scattered console.log() and console.error() calls with a centralized system.
 *
 * Benefits:
 * - Consistent log format across the app
 * - Easy to disable/enable logging by environment
 * - Structured data for easier debugging
 * - Can be extended to send logs to monitoring services
 * - Performance insights with timestamps
 * - Context-rich error reporting
 *
 * Usage:
 * ```typescript
 * import { logger } from '@/utils/logger';
 *
 * logger.info('User logged in', { userId: '123', email: 'user@example.com' });
 * logger.error('API call failed', { endpoint: '/api/calendars', error });
 * logger.debug('Rendering calendar', { calendarId: 'cal-456' });
 * ```
 */

/**
 * Log level enumeration
 */
export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR'
}

/**
 * Log entry interface
 */
interface LogEntry {
  level: LogLevel;
  message: string;
  timestamp: string;
  data?: Record<string, unknown>;
  stack?: string;
}

/**
 * Logger configuration
 */
interface LoggerConfig {
  /** Whether logging is enabled */
  enabled: boolean;
  /** Minimum log level to display */
  minLevel: LogLevel;
  /** Whether to include timestamps */
  includeTimestamp: boolean;
  /** Whether to include stack traces for errors */
  includeStackTrace: boolean;
}

/**
 * Default logger configuration
 * - Enabled in development, disabled in production (can be overridden)
 * - INFO level and above in production
 * - DEBUG level and above in development
 */
const defaultConfig: LoggerConfig = {
  enabled: !import.meta.env.PROD,
  minLevel: import.meta.env.PROD ? LogLevel.INFO : LogLevel.DEBUG,
  includeTimestamp: true,
  includeStackTrace: true
};

/**
 * Current logger configuration
 */
let config: LoggerConfig = { ...defaultConfig };

/**
 * Log level priority (for filtering)
 */
const levelPriority: Record<LogLevel, number> = {
  [LogLevel.DEBUG]: 0,
  [LogLevel.INFO]: 1,
  [LogLevel.WARN]: 2,
  [LogLevel.ERROR]: 3
};

/**
 * Format log entry for console output
 */
function formatLogEntry(entry: LogEntry): string[] {
  const parts: string[] = [];

  // Timestamp
  if (config.includeTimestamp) {
    parts.push(`[${entry.timestamp}]`);
  }

  // Log level with color coding
  parts.push(`[${entry.level}]`);

  // Message
  parts.push(entry.message);

  return parts;
}

/**
 * Get console method for log level
 */
function getConsoleMethod(level: LogLevel): (...args: unknown[]) => void {
  switch (level) {
    case LogLevel.DEBUG:
      return console.debug;
    case LogLevel.INFO:
      return console.info;
    case LogLevel.WARN:
      return console.warn;
    case LogLevel.ERROR:
      return console.error;
    default:
      return console.log;
  }
}

/**
 * Get color for log level (for browser console)
 */
function getLogLevelColor(level: LogLevel): string {
  switch (level) {
    case LogLevel.DEBUG:
      return 'color: #9E9E9E'; // Gray
    case LogLevel.INFO:
      return 'color: #2196F3'; // Blue
    case LogLevel.WARN:
      return 'color: #FF9800'; // Orange
    case LogLevel.ERROR:
      return 'color: #F44336'; // Red
    default:
      return 'color: #000000'; // Black
  }
}

/**
 * Internal log function
 */
function log(level: LogLevel, message: string, data?: Record<string, unknown>): void {
  // Check if logging is enabled
  if (!config.enabled) {
    return;
  }

  // Check log level
  if (levelPriority[level] < levelPriority[config.minLevel]) {
    return;
  }

  // Create log entry
  const entry: LogEntry = {
    level,
    message,
    timestamp: new Date().toISOString(),
    data
  };

  // Add stack trace for errors
  if (level === LogLevel.ERROR && config.includeStackTrace) {
    entry.stack = new Error().stack;
  }

  // Format and output
  const formattedParts = formatLogEntry(entry);
  const consoleMethod = getConsoleMethod(level);

  // Use styled console output in browser
  if (typeof window !== 'undefined') {
    consoleMethod(
      `%c${formattedParts[0]} %c${formattedParts[1]} %c${formattedParts[2]}`,
      'color: #9E9E9E',
      getLogLevelColor(level),
      'color: inherit'
    );

    // Log data object if provided
    if (data && Object.keys(data).length > 0) {
      consoleMethod('Data:', data);
    }

    // Log stack trace if available
    if (entry.stack) {
      consoleMethod('Stack:', entry.stack);
    }
  } else {
    // Plain output for non-browser environments
    consoleMethod(...formattedParts);
    if (data) {
      consoleMethod('Data:', data);
    }
  }
}

/**
 * Logger class with public API
 */
class Logger {
  /**
   * Log debug message (lowest priority)
   * Use for detailed debugging information
   */
  debug(message: string, data?: Record<string, unknown>): void {
    log(LogLevel.DEBUG, message, data);
  }

  /**
   * Log info message (normal priority)
   * Use for general informational messages
   */
  info(message: string, data?: Record<string, unknown>): void {
    log(LogLevel.INFO, message, data);
  }

  /**
   * Log warning message (high priority)
   * Use for potentially harmful situations
   */
  warn(message: string, data?: Record<string, unknown>): void {
    log(LogLevel.WARN, message, data);
  }

  /**
   * Log error message (highest priority)
   * Use for error events that might still allow the app to continue
   */
  error(message: string, data?: Record<string, unknown>): void {
    log(LogLevel.ERROR, message, data);
  }

  /**
   * Update logger configuration
   */
  configure(newConfig: Partial<LoggerConfig>): void {
    config = { ...config, ...newConfig };
  }

  /**
   * Get current logger configuration
   */
  getConfig(): LoggerConfig {
    return { ...config };
  }

  /**
   * Reset logger configuration to defaults
   */
  resetConfig(): void {
    config = { ...defaultConfig };
  }

  /**
   * Enable logging
   */
  enable(): void {
    config.enabled = true;
  }

  /**
   * Disable logging
   */
  disable(): void {
    config.enabled = false;
  }
}

/**
 * Singleton logger instance
 */
export const logger = new Logger();

/**
 * Default export for convenience
 */
export default logger;
